File: content/snippets/go/exercises/advanced/parallel_aggregator.go
package main

import (
	"fmt"
	"strings"
	"sync"
)

type result struct {
	endpoint string
	ok       bool
	payload  string
}

func fetch(ep string) result {
	if strings.Contains(ep, "fail") {
		return result{ep, false, "timeout"}
	}
	return result{ep, true, fmt.Sprintf("{\"endpoint\":\"%s\"}", ep)}
}

func main() {
	endpoints := []string{"users", "orders", "fail-metrics"}
	ch := make(chan result, len(endpoints))
	var wg sync.WaitGroup

	for _, ep := range endpoints {
		wg.Add(1)
		go func(e string) {
			defer wg.Done()
			ch <- fetch(e)
		}(ep)
	}

	wg.Wait()
	close(ch)

	var data []string
	var errors []string
	for r := range ch {
		if r.ok {
			data = append(data, r.payload)
		} else {
			errors = append(errors, r.endpoint+": "+r.payload)
		}
	}

	fmt.Printf("data=%d errors=%d\n", len(data), len(errors))
}

File: content/snippets/python/exercises/advanced/parallel_aggregator.py
import asyncio

async def fetch(endpoint: str) -> tuple[str, bool, str]:
    await asyncio.sleep(0.05)
    if "fail" in endpoint:
        return endpoint, False, "timeout"
    return endpoint, True, f"{{\"endpoint\": \"{endpoint}\"}}"

async def main() -> None:
    endpoints = ["users", "orders", "fail-metrics"]
    results = await asyncio.gather(*(fetch(ep) for ep in endpoints), return_exceptions=False)

    data = [payload for _, ok, payload in results if ok]
    errors = [f"{ep}: {payload}" for ep, ok, payload in results if not ok]

    print({"data": data, "errors": errors})

asyncio.run(main())
